python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -699,6 +700,125 @@ def ngroup(self, ascending=True):
+                [*self.obj._index._columns]  # type: ignore
+            )
+        except ValueError:
+            raise NotImplementedError(
+                "Sorry groupby.sample with multiindex not implemented"
+            )
+        group_offsets = np.asarray(group_offsets).astype(np.int32)
+        size_per_group = np.diff(group_offsets)
+        if n is not None:
+            samples_per_group = np.broadcast_to(
+                np.int32(n), size_per_group.shape
+            )
+            if not replace and (minsize := size_per_group.min()) < n:
+                raise ValueError(
+                    f"Cannot sample {n=} without replacement, "
+                    f"smallest group is {minsize}"
+                )
+        else:
+            # Pandas uses round-to-nearest, ties to even to
+            # pick sample sizes for the fractional case (unlike IEEE
+            # which is round-to-nearest, ties to sgn(x) * inf).
+            samples_per_group = np.round(
+                size_per_group * frac, decimals=0
+            ).astype(np.int32)
+        if replace:
+            # Use numpy to do all-at-once generation of the
+            # We can use numpy in this case, which is fast for lots of
+            # groups and large samples
+            lo = 0
+            hi = np.repeat(size_per_group, samples_per_group)
+            rng = np.random.default_rng(seed=random_state)
+            # Would be nice to use cupy here, but their rng.integers
+            # interface doesn't take array lo and hi arguments.
+            indices = rng.integers(lo, hi, dtype=np.int32).reshape(-1)
+            indices += np.repeat(group_offsets[:-1], samples_per_group)
+        else:
+            # Approach: do a segmented argsort of the index array and take
+            # the first samples_per_group entries from sorted array.
+            if len(size_per_group) < 500:
+                # Empirically shuffling with cupy is faster at this scale
+                indices = cp.asarray(index.data_array_view(mode="read"))
+                rs = cp.random.get_random_state()
+                rs.seed(seed=random_state)
+                for off, size in zip(group_offsets, size_per_group):
+                    rs.shuffle(indices[off : off + size])
+            else:
+                rng = cp.random.default_rng(seed=random_state)
+                offsets_col = as_column(group_offsets)
+                values = [index]
+                keys = [as_column(rng.random(size=len(index)))]
+                (indices,) = segmented_sort_by_key(
+                    values, keys, offsets_col, [], []
+                )
+                indices = cp.asarray(indices.data_array_view(mode="read"))
+            # Which indices are we going to want?
+            mask = np.zeros(len(index), dtype=bool)
+            for offset, sample_size in zip(group_offsets, samples_per_group):
+                mask[offset : offset + sample_size] = True
+            indices = indices[mask]
```

Comment:
I would like to test an alternative here, which might be more efficient with cupy (no raw loops, just array operators). I don't know how it will compare to the CPU NumPy implementation. Here's the sketch. You might benchmark it for various group sizes / sample sizes.
```python
import numpy as np

num_groups = 100_000
items_per_group = 10

group_offsets = np.arange(0, num_groups*items_per_group, items_per_group)
samples_per_group = np.repeat(items_per_group//3, num_groups)
indices = np.arange(0, num_groups*items_per_group)

def take_mask(indices, group_offsets, samples_per_group):
    # Which indices are we going to want?
    mask = np.zeros(len(indices), dtype=bool)
    for offset, sample_size in zip(group_offsets, samples_per_group):
        mask[offset : offset + sample_size] = True
    return indices[mask]

def take_offsets(indices, group_offsets, samples_per_group):
    to_take = np.arange(np.sum(samples_per_group))
    exclusive_scan = np.empty(len(samples_per_group), dtype=np.int_)
    exclusive_scan[0] = 0
    exclusive_scan[1:] = np.cumsum(samples_per_group[:-1])
    to_take += np.repeat(group_offsets - exclusive_scan, samples_per_group)
    return indices[to_take]

mask_result = take_mask(indices, group_offsets, samples_per_group))
offsets_result = take_offsets(indices, group_offsets, samples_per_group))
```

Reply:
I'll have a go on monday. Also, I realise this in a broken state as soon as I added non-default indices to the dataframe (so I'll untangle that too).